sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CNFNormalizationSuite.scala

```diff
@@ -0,0 +1,120 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.expressions
+
+import scala.collection.mutable.ArrayBuffer
+
+import org.apache.spark.SparkFunSuite
+import org.apache.spark.sql.catalyst.plans.logical._
+import org.apache.spark.sql.catalyst.dsl.plans._
+import org.apache.spark.sql.catalyst.dsl.expressions._
+
+class CNFNormalizationSuite extends SparkFunSuite with PredicateHelper {
+
+  val testRelation = LocalRelation('a.int, 'b.int, 'c.int, 'd.int, 'e.int)
+
+  // Take in expression without And, normalize to its leftmost representation to be comparable
+  private def normalizeOrExpression(expression: Expression): Expression = {
+    val atoms = ArrayBuffer.empty[Expression]
+    expression.foreachUp {
+      case Or(l: Or, r: Or) => // do nothing
+      case Or(l, r: Or) => atoms += l
+      case Or(l: Or, r) => atoms += r
+      case Or(l, r) => atoms += l; atoms += r
+      case _ => // do nothing
+    }
+    if (!expression.isInstanceOf[Or]) {
+      atoms += expression
+    }
+    atoms.sortBy(_.toString).reduce(Or)
+  }
+
+  private def checkCondition(input: Expression, expected: Expression): Unit = {
+    val actual = testRelation.where(input).analyze
+    val correctAnswer = testRelation.where(expected).analyze
+
+    val resultFilterExpression = actual.collectFirst { case f: Filter => f.condition }.get
+    val expectedFilterExpression = correctAnswer.collectFirst { case f: Filter => f.condition }.get
+
+    val exprs = splitConjunctivePredicates(cnfNormalization(resultFilterExpression)).
+      map(normalizeOrExpression).sortBy(_.toString)
```

Comment:
I think I still need `reduce(Or)` because after `flatMap` I cannot sort them. therefore, (a || b || c) && (a || b || d) is still not equal to (a || b || d) && (a || b || c).